dali/python/nvidia/dali/pipeline.py

```diff
@@ -1884,6 +1885,98 @@ def create_pipeline(*args, **kwargs):
     return actual_decorator(fn) if fn else actual_decorator
 
 
+# Callable preserving a signature
+_F = TypeVar("_F", bound=Callable[..., Any])
+
+
+def do_not_convert(func: _F = None) -> _F:
+    """Decorator that suppresses the conversion of a function by AutoGraph.
+
+    In conditional mode, DALI uses a fork of
+    `TensorFlow's AutoGraph <https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/autograph/g3doc/reference/index.md>`_
+    to transform the code, enabling us to rewrite and detect the ``if`` statements, so they can be
+    used in processing the DALI pipeline.
+
+    When used with :meth:`external source <nvidia.dali.fn.external_source>` in parallel mode
+    (``parallel=True``), this may interfere with the serialization of the provided ``source``
+    parameter. To prevent this, functions that are used to create the ``source`` parameter,
+    should be decorated with :meth:`@do_not_convert <nvidia.dali.pipeline.do_not_convert>`.
+
+    The AutoGraph conversion is applied to any top-level function or method called within the
+    pipeline definition (as well as the pipeline definition itself).
+    When a function is converted, all functions defined within its syntactical scope are also
+    converted.
+
+    To stop a function from being converted, its top-level encompassing function must be marked
+    with this decorator. This may sometimes require refactoring the function to outer scope.
+
+    .. note::
+       Only functions that do not process :class:`DataNode` (so do not use DALI operators)
+       should be marked with this decorator.
+
+    For example::
+
+        from nvidia.dali import pipeline_def, fn
+
+        @pipeline_def(enable_conditionals=True)
+        def pipe():
+
+            def source_factory(size):
+                def source_fun(sample_info):
+                    return np.full(size, sample_info.iter_idx)
+                return source_fun
```

Comment:
Can't we just do:
```suggestion
            @do_not_convert
            def source_factory(size):
                def source_fun(sample_info):
                    return np.full(size, sample_info.iter_idx)
                return source_fun
```
?
This would be much less invasive.

Reply:
> When a function is converted, all functions defined within its syntactical scope are also converted.

Adding a decorator doesn't stop this process, we must prevent a conversion when a function from external scope is called, as those from this scope, were already converted.

Reply:
...and can't we tweak autograph so it doesn't touch those?